### PR TITLE
Ge signa updates

### PR DESCRIPTION
**Pull request by pwadhwa351**

### CODE REVIEW

File: src/include/stir/listmode/CListRecordGESigna.h [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    Copyright (C) 2003-2011 Hammersmith Imanet Ltd (CListRecordECAT.h)
+    Copyright (C) 2013 University College London (major mods for GE Dimension data)
+    Copyright (C) 2017-2018 University of Leeds
+*/
+/*!
+  \file
+  \ingroup listmode
+  \brief Classes for listmode records of GE Dimension console data
+
+  This file is based on GE proprietary information and can therefore not be distributed outside UCL
```
Comment (KrisThielemans) [inline src/include/stir/listmode/CListRecordGESigna.h]: delete

File: src/include/stir/listmode/CListRecordGESigna.h [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    /* RESERVED     = 0x4 to 0xE */
+    /* 0xE is temporary taken here to mark end of it. */
+    END_LIST_EVT = 0xE,
+    SINGLE_EVT      = 0xF
+};
+
+
+//! Class for storing and using a coincidence event from a GE Dimension listmode file
+/*! \ingroup listmode
+  This class cannot have virtual functions, as it needs to just store the data 4 bytes for CListRecordGESigna to work.
```
Comment (KrisThielemans) [inline src/include/stir/listmode/CListRecordGESigna.h]: 4 bytes is incorrect

File: src/include/stir/listmode/CListRecordGESigna.h [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    boost::uint16_t hiXtalTransAxID:10;  /* High Crystal Trans-Axial Id */
+    boost::uint16_t loXtalAxialID:6;     /* Low Crystal Axial Id */
+    boost::uint16_t loXtalTransAxID:10;  /* Low Crystal Trans-Axial Id */
+#endif
+}; /*-coincidence event*/
+
+
+//! A class for storing and using a timing 'event' from a GE Signa PET/MR listmode file
+/*! \ingroup listmode
+  This class cannot have virtual functions, as it needs to just store the data 8 bytes for CListRecordGESigna to work.
```
Comment (KrisThielemans) [inline src/include/stir/listmode/CListRecordGESigna.h]: 8 bytes is incorrect

File: src/include/stir/listmode/CListRecordGESigna.h [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+//! A class for a general element (or "record") of a GE Dimension listmode file
+/*! \ingroup listmode
+  All types of records are stored in a (private) union with the "basic" classes such as CListEventDataGESigna.
+  This class essentially just forwards the work to the "basic" classes.
+
+  A complication for GE Dimension data is that not all events are the same size:
+  coincidence events are 4 bytes, and others are 8 bytes. 
+
+  \todo Currently we always assume the data is from a DSTE. We should really read this from the RDF header.
```
Comment (KrisThielemans) [inline src/include/stir/listmode/CListRecordGESigna.h]: delete

File: src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h [hunk 5]
```diff
@@ -0,0 +1,143 @@
+/*
+  Copyright (C) 2000-2007, Hammersmith Imanet Ltd
+  Copyright (C) 2013-2014 University College London
```
Comment (KrisThielemans) [inline src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h]: correct

File: src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+using std::string;
+#endif
+
+START_NAMESPACE_STIR
+
+/*!
+  \ingroup recon_buildblock
+  \ingroup GE
+  \brief A BinNormalisation class that gets the normalisation factors from
+  an GEHDF5 3D normalisation file. Note that you have to point it to the
```
Comment (KrisThielemans) [inline src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h]: the comment about the Interfile header needs to be deleted

File: src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+  \par Parsing example
+  \verbatim
+  Bin Normalisation type := from GE HDF5
+  Bin Normalisation From GEHDF5:=
+  normalisation filename:= myfile.hn
+
+  ; next keywords can be used to switch off some of the normalisation components
+  ; do not use unless you know why
+  ; use_gaps:=1
```
Comment (KrisThielemans) [inline src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h]: delete

File: src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  // TODO move to Scanner
+  int num_axial_blocks_per_singles_unit;
+  shared_ptr<ProjDataInfo> proj_data_info_ptr;
+  ProjDataInfoCylindricalNoArcCorr const * proj_data_info_cyl_ptr;
+  shared_ptr<ProjDataInfoCylindricalNoArcCorr> proj_data_info_cyl_uncompressed_ptr;
+  int span;
+  int mash;
+  int num_blocks_per_singles_unit;
+
+  bool _use_gaps;
```
Comment (KrisThielemans) [inline src/include/stir/recon_buildblock/BinNormalisationFromGEHDF5.h]: delete

File: src/listmode_buildblock/CListModeDataGESigna.cxx [hunk 9]
```diff
@@ -0,0 +1,131 @@
+/*
+    Copyright (C) 2013 University College London
```
Comment (KrisThielemans) [inline src/listmode_buildblock/CListModeDataGESigna.cxx]: correct?

File: src/listmode_buildblock/CListModeDataGESigna.cxx [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+*/
+/*!
+  \file
+  \ingroup listmode  
+  \brief Implementation of class stir::CListModeDataGESigna
+    
+  \author Kris Thielemans
```
Comment (KrisThielemans) [inline src/listmode_buildblock/CListModeDataGESigna.cxx]: correct?

File: src/listmode_buildblock/CListModeDataGESigna.cxx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+#endif
+
+  //! \todo N.E: Probably can do without the HDF5Wrapper here.
+  HDF5Wrapper inputFile(listmode_filename);
+  CListModeData::scanner_sptr = inputFile.get_scanner_sptr();
+  
+  //! \todo N.E: Remove hard-coded sizes;
+  current_lm_data_ptr.
+  reset(
+        new InputStreamWithRecordsFromHDF5<CListRecordT>(listmode_filename,
```
Comment (KrisThielemans) [inline src/listmode_buildblock/CListModeDataGESigna.cxx]: are we sure about this? I thought they're always 6 bytes? if you don't know, put a todo comment

File: src/utilities/conv_GEHDF5_to_interfile.cxx [hunk 12]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (KrisThielemans) [inline src/utilities/conv_GEHDF5_to_interfile.cxx]: delete this file. `stir_math -s` will do the same

File: src/utilities/read_geo_factors.cxx [hunk 13]
```diff
@@ -0,0 +1,132 @@
+/*
```
Comment (KrisThielemans) [inline src/utilities/read_geo_factors.cxx]: do we need this file once the `BinNormalisation` is sorted?  If we do (or you want to keep it), rename it to `read_GEHEAD5_geo_factors`

File: src/utilities/read_write_singles.cxx [hunk 14]
```diff
@@ -0,0 +1,255 @@
+/*!
```
Comment (KrisThielemans) [inline src/utilities/read_write_singles.cxx]: if you want to keep this file, rename and clean

**Comment by KrisThielemans** [comment]
All Travis jobs are failing because it attempts to compile buildblock/ProjDataFromHDF5.cxx:
```
In file included from /home/travis/build/UCL/STIR/src/include/stir/ProjDataFromHDF5.h:38:0,
                 from /home/travis/build/UCL/STIR/src/buildblock/ProjDataFromHDF5.cxx:34:
/home/travis/build/UCL/STIR/src/include/stir/IO/HDF5Wrapper.h:32:19: fatal error: H5Cpp.h: No such file or directory
 #include "H5Cpp.h"
                   ^
compilation terminated.
```
This is weird as `IO/CMakeLists.txt` only includes it `if (HDF5_FOUND)`, which is as it should and CMake says
```
Could NOT find HDF5 (missing: HDF5_LIBRARIES HDF5_INCLUDE_DIRS CXX) 
```
example job [here](https://travis-ci.org/UCL/STIR/jobs/599265269)

File: src/buildblock/CMakeLists.txt [hunk 15]
```diff
@@ -13,8 +13,9 @@ set(${dir_LIB_SOURCES}
   ProjDataInfoCylindricalArcCorr 
   ProjDataInfoCylindricalNoArcCorr 
   ArcCorrection 
-  ProjDataFromStream 
-  ProjDataGEAdvance 
+  ProjDataFromHDF5
```
Comment (KrisThielemans) [inline src/buildblock/CMakeLists.txt]: This needs to be deleted  (it's already below, but there properly guarded)

File: CMakeLists.txt [hunk 16]
```diff
@@ -84,6 +84,7 @@ option(DISABLE_LLN_MATRIX "disable use of LLN library" OFF)
 option(DISABLE_ITK "disable use of ITK library" OFF)
 option(DISABLE_AVW "disable use of AVW library" OFF)
 option(DISABLE_RDF "disable use of GE RDF library" OFF)
+option(DISABLE_HDF5 “disable use of HDF5 libraries” OFF)
```
Comment (KrisThielemans) [inline CMakeLists.txt]: it's not happy with this. I expect because the wrong quotes are used.

File: CMakeLists.txt [hunk 17]
```diff
@@ -119,6 +120,10 @@ if(NOT DISABLE_RDF)
   find_package(RDF)
 endif()
 
+if(NOT DISABLE_HDF5_SUPPORT)
```
Comment (KrisThielemans) [inline CMakeLists.txt]: just `DISABLE_HDF5`

File: src/CMakeLists.txt [hunk 18]
```diff
@@ -68,6 +70,16 @@ else()
   message(STATUS "RDF support disabled.")
 endif()
 
+if (HDF5_FOUND)
+  set(HAVE_HDF5 ON)
+  message(STATUS "HDF5 support enabled.")
+  add_definitions(-D HAVE_HDF5)
```
[the remark this answers is not in the document]
Comment (pwadhwa351) [reply src/CMakeLists.txt]: I adapted the code as above, I am getting following error during build:
Undefined symbols for architecture x86_64:
  "stir::SinglesRatesFromGEHDF5::read_singles_from_listmode_file(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      _main in construct_randoms_from_GEsingles.cxx.o
  "stir::SinglesRatesFromGEHDF5::SinglesRatesFromGEHDF5()", referenced from:
      _main in construct_randoms_from_GEsingles.cxx.o
  "stir::SinglesRatesFromGEHDF5::get_singles_rate(stir::DetectionPosition<unsigned int> const&, double, double) const", referenced from:
      _main in construct_randoms_from_GEsingles.cxx.o
  "stir::SinglesRatesFromGEHDF5::get_num_time_slices() const", referenced from:
      _main in construct_randoms_from_GEsingles.cxx.o
  "vtable for stir::SinglesRatesFromGEHDF5", referenced from:
      stir::SinglesRatesFromGEHDF5::~SinglesRatesFromGEHDF5() in construct_randoms_from_GEsingles.cxx.o
[ 42%] Built target conv_gipl_to_interfile
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

File: src/include/stir/listmode/CListModeData.h [hunk 19]
```diff
@@ -243,6 +243,7 @@ class CListModeData : public ExamData
   void set_proj_data_info_sptr(shared_ptr<ProjDataInfo>);
   //! Has to be set by the derived class
   //  shared_ptr<ExamInfo> exam_info_sptr;
+  shared_ptr<Scanner> scanner_sptr;
```
Comment (KrisThielemans) [inline src/include/stir/listmode/CListModeData.h]: this has to be removed. It was here before, but is now obsolete (as we have one in `proj_data_info_sptr`

File: src/CMakeLists.txt (hunk 18 above)
[the remark this answers is not in the document]
Comment (KrisThielemans) [reply src/CMakeLists.txt]: not sure. I guess you haven't committed this change (which is fine as it doesn't work!) as I can't see it.

What did you change then? I only suggested to remove the line `add_definitions(-D HAVE_HDF5)` (replacing it with the change in the `.h.in`). What you're getting is a linker error which seems to say that `SinglesRatesFromGEHDF5.cxx` isn't compiled, which could mean that `HDF5_FOUND` is false.

**Comment by KrisThielemans** [comment]
Note: some travis jobs were failing due to a problem on `master` (now fixed). sorry about that. I'm afraid you'll have to merge `master` back in again.

**Comment by KrisThielemans** [comment]
By the way, could you give maintainers push permission?
https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

I can then commit minor suggestions myself to save you time.

**Comment by pwadhwa351** [comment]
The maintainers now have push permissions hopefully!
________________________________

**Comment by KrisThielemans** [comment]
ok. thanks . Shall I quickly try to resolve the CMake stuff? I'll leave the rest for you!

**Comment by pwadhwa351** [comment]
Yes, that should be helpful thanks a lot!
________________________________

**Comment by KrisThielemans** [comment]
Travis ok now. I'll now enable a few jobs with HDF5 on such that it'll actually compile this code.

**Comment by KrisThielemans** [comment]
some travis failures because this code requires C++-11: HDF5Wrapper uses `std::array`. Unfortunately we've scheduled C++-11 for v5.0, #99. We could use `boost::array` when `std::array` doesn't exist.

**Comment by KrisThielemans** [comment]
I've decided not to work-around the C++-11 problem. There were other "fixes" needed (non-static member initialisations), which just would have made the code uglier. I therefore tell cmake to use C++-11 if HDF5 is found. Not so nice, but it's who needs such an old C++ anyway...

This caused a problem with clang that it warns about auto_ptr if C++-11 is enabled (somewhat ironically, as in clang-5, auto_ptr was broken). I've changed the Travis file to add `-Wno-deprecated-declarations` for this case. Somewhat optimistically, I've then enabled the clang-6 check as well (which had the same problem).

Hopefully, Travis will be green after this! Appveyor was already (but it doesn't test the HDF5 code at the moment).

@pwadhwa351, while I'm messing about, it's best to `git pull` before you commit, or even `git pull --rebase` after you've committed when you cannot push (but don't do the latter if you have pushed your own commits elsewhere).

File: src/IO/HDF5Wrapper.cxx [hunk 20]
```diff
@@ -0,0 +1,543 @@
...
+    //  // TODO error checking
+    return Succeeded::yes;
+}
+
+// Developed for Singles
+Succeeded HDF5Wrapper::get_dataspace(const unsigned int current_id,
+                                     Array<1, unsigned int>& output)
+{
+    std::ostringstream datasetname;
+    datasetname << m_address << current_id;
```
[the remark this answers is not in the document]
Comment (pwadhwa351) [reply src/IO/HDF5Wrapper.cxx]: When this function is used to get the dataspace in class SinglesRatesFromGEHDF5.cxx, this is used in conjunction with the function initialise_singles_data() which also fills up m_address to get the respective position in HDF5 file. Then the dataspace information is extracted for that particular address so its needed to here.

[the remark this answers is not in the document]
Comment (KrisThielemans) [reply src/IO/HDF5Wrapper.cxx]: very opaque to the user. Any reason why you can't just pass m_address, I guess same as the `offset` above. then this function becomes independent of other stuff

File: src/include/stir/IO/GESignaListmodeInputFileFormat.h [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+      }
+  }
+ public:
+  virtual unique_ptr<data_type>
+    read_from_file(std::istream& input) const
+  {
+    warning("read_from_file for GESigna listmode data with istream not implemented %s:%s. Sorry",
+	  __FILE__, __LINE__);
+    return
+      std::auto_ptr<data_type>
```
[the remark this answers is not in the document]
Comment (pwadhwa351) [reply src/include/stir/IO/GESignaListmodeInputFileFormat.h]: Using unique_ptr is giving me this error:
/Users/palak/STIR_upstream/STIR/src/include/stir/IO/GESignaListmodeInputFileFormat.h:103: error: ambiguous conversion for functional-style cast from 'int' to 'unique_ptr<data_type>' (aka 'unique_ptr<stir::CListModeData>')
      unique_ptr<data_type>
      ^~~~~~~~~~~~~~~~~~~~~
Apologies for my limited knowledge!

File: src/utilities/read_geo_factors.cxx (hunk 13 above)
Comment (pwadhwa351) [reply src/utilities/read_geo_factors.cxx]: No this file won't be there once it is shifted to BinNormalisationFromGEHDF5

File: src/data_buildblock/SinglesRatesFromGEHDF5.cxx [hunk 22]
```diff
@@ -47,9 +47,9 @@ using std::ios;
 
 
 START_NAMESPACE_STIR
-//PW Need to change the registered name
+//PW todo Need to change the registered name
```
Comment (KrisThielemans) [inline src/data_buildblock/SinglesRatesFromGEHDF5.cxx]: can remove the todo now

File: src/include/stir/IO/HDF5Wrapper.h [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+{
+public:
+
+    static bool check_GE_signature(const std::string filename);
+
+    explicit HDF5Wrapper();
+
+    explicit HDF5Wrapper(const std::string& filename);
+
+    Succeeded open(const std::string& filename);
```
[the remark this answers is not in the document]
Comment (pwadhwa351) [reply src/include/stir/IO/HDF5Wrapper.h]: All the other requested changes have been included and accounted for except this as most of these functions are used outside the scope.

**Comment by KrisThielemans** [comment]
Do you want to merge this on top of the release branch before merging to master? Might be easiest. Don't do the merges yet. first let me know.

**Comment by KrisThielemans** [comment]
I've changed this to sit on top of the `release_4` branch. We'll merge this PR there, hopefully adding it to STIR 4.1, before going to `master`.

**Comment by KrisThielemans** [comment]
@pwadhwa351 I've added something in the release_4.1.htm. Please check. I've also renamed `ProjDataFromHDF5` to `ProjDataGEHDF5`. However, this isn't used in `ProjData.cxx`. This confuses me a bit. Maybe it got lost in one of the merges? Or were you reading these files using specific functionality? (I guess so).

File: src/listmode_buildblock/CListModeDataGESigna.cxx (hunk 11 above)
Comment (KrisThielemans) [reply src/listmode_buildblock/CListModeDataGESigna.cxx:105]: Palak did confirm this is ok

**Comment by KrisThielemans** [comment]
Things remaining (won't fix this now as I have no data/time to test)
- various hard-wired numbers for the Signa
-  `SinglesRatesFromGEHDF5` 
    - expose GEHDF5Wrapper.get_exam_info_sptr()
    - `construct_randoms_from_GEsingles.cxx` needs to reorder data. This should be moved to this class
- `GESignaListmodeInputFileFormat` only checks if it's an HDF5 file
- `ProjData` should call `ProjDataGEHDF5`

**Pull request merged**